flask_restful/__init__.py

```diff
@@ -287,33 +297,86 @@ def handle_error(self, e):
-            code = e.code
-            default_data = {
-                'message': getattr(e, 'description', http_status_message(code))
-            }
-            headers = e.get_response().headers
+        for typecheck, handler in self.errorhandlers:
+            if isinstance(e, typecheck):
+                try:
+                    # unpack(None) -> (None, _, _)
+                    data, code, headers = unpack(handler(e))
+                    if data:
+                        break
+                except:
+                    # log exception and return http 500
+                    current_app.log_exception(sys.exc_info())
+                    data, code, headers = self.default_http_exception_handler(
+                        InternalServerError('Exception in error handler!'))
+                    break
         else:
-            code = 500
-            default_data = {
-                'message': http_status_message(code),
-            }
+            # `default_exception_handler` is always at the bottom, even if no
+            # error handler was registered
+            data, code, headers = self.default_exception_handler(e)
+
+        return self.make_response(data, code, headers)
+
+    def default_exception_handler(self, e):
+        """Default response for an unhandled Exception
 
+        :param e: the exception raised while handling the request
+        :type e: Exception
+        :return: (data, code, headers) tuple
+        """
+        code = 500
+        default_data = { 'message': http_status_message(code), }
         data = getattr(e, 'data', default_data)
+        return data, code, Headers()
 
-        if code >= 500:
-            exc_info = sys.exc_info()
-            if exc_info[1] is None:
-                exc_info = None
-            current_app.log_exception(exc_info)
+    def default_http_exception_handler(self, e):
+        """Default response for a HTTPException
+
+        :param e: the exception raised while handling the request
+        :type e: werkzeug.exceptions.HTTPException
+        :return: (data, code, headers) tuple
+        """
+        code = e.code
+        default_data = { 'message': getattr(e, 'description', http_status_message(code)), }
+        data = getattr(e, 'data', default_data)
+        return data, code, e.get_response().headers
+
+    def custom_errors_handlers(self, e):
+        """Process custom error handlers passed by `errors` argument
 
+        :param e: the exception raised while handling the request
+        :type e: Exception
+        :return: (data, code, headers) tuple or None if no match was found
+        """
+        error_cls_name = type(e).__name__
+        if error_cls_name in self.errors:
+            if isinstance(e, HTTPException):
+                data, code, headers = self.default_http_exception_handler(e)
+            else:
+                data, code, headers = self.default_exception_handler(e)
+            custom_data = self.errors.get(error_cls_name, {})
```

Comment:
With this new errorhandler system, I don't see a reason to keep the old system of initializing the `Api` object with a dict of custom response data mapped to exception names.